jlm/rvsdg/Phi.hpp

```diff
@@ -0,0 +1,383 @@
+   *
+   * Returns all fixpoint variable descriptions.
+   */
+  [[nodiscard]] std::vector<FixVar>
+  GetFixVars() const noexcept;
+
+  /**
+   * \brief Tries to map region argument to fixpoint variable.
+   *
+   * \param argument
+   *   The argument of the region to be mapped.
+   *
+   * \returns
+   *   Fixpoint variable description corresponding to \p argument,
+   *   or \p std::nullopt.
+   *
+   * \pre
+   *   \p argument must be an argument of the subregion contained in
+   *   the phi node.
+   *
+   * Checks whether the given argument corresponds to a fixpoint
+   * variable and returns its description in that case. All
+   * arguments of a phi region are either context or fixpoint
+   * variables, see \ref MapArgumentContextVar.
+   */
+  [[nodiscard]] std::optional<FixVar>
+  MapArgumentFixVar(const rvsdg::output & argument) const noexcept;
+
+  /**
+   * \brief Maps region result to fixpoint variable.
+   *
+   * \param result
+   *   The result of the region to be mapped.
+   *
+   * \returns
+   *   Fixpoint variable description corresponding to \p input.
+   *
+   * \pre
+   *   \p result must be a result of the subregion contained in
+   *   the phi node.
+   *
+   * Maps result of the region to a fixpoint variable.
+   */
+  [[nodiscard]] FixVar
+  MapResultFixVar(const rvsdg::input & result) const noexcept;
+
+  /**
+   * \brief Maps output to fixpoint variable.
+   *
+   * \param output
+   *   The output of the phi node to be mapped.
+   *
+   * \returns
+   *   Fixpoint variable description corresponding to \p output.
+   *
+   * \pre
+   *   \p output must be an output of this phi node.
+   *
+   * Maps output of the phi node to a fixpoint variable.
+   */
+  [[nodiscard]] FixVar
+  MapOutputFixVar(const rvsdg::output & output) const noexcept;
+
+  /**
+   * \brief Maps region argument to its function.
+   *
+   * \param argument
+   *   The argument of the phi node to be mapped.
+   *
+   * \returns
+   *   Either the fixpoint variable description or the
+   *   context variable description, depending on the
+   *   designation of the argument.
+   *
+   * \pre
+   *   \p argument must be an argument of the region within this phi node.
+   *
+   * Maps output of the phi node to either a fixpoint or context variable.
+   */
+  [[nodiscard]] std::variant<FixVar, ContextVar>
+  MapArgument(const rvsdg::output & argument) const noexcept;
+
+  rvsdg::Region *
+  subregion() const noexcept
+  {
+    return StructuralNode::subregion(0);
+  }
+
+  virtual PhiNode *
+  copy(rvsdg::Region * region, rvsdg::SubstitutionMap & smap) const override;
+
+  /**
+   * Extracts all lambda nodes from a phi node.
+   *
+   * The function is capable of handling nested phi nodes.
+   *
+   * @param phiNode The phi node from which the lambda nodes should be extracted.
+   * @return A vector of lambda nodes.
+   */
+  static std::vector<rvsdg::LambdaNode *>
+  ExtractLambdaNodes(const PhiNode & phiNode);
+};
+
+/* phi builder class */
+
```

Comment:
Can be removed or replaced with proper documentation

Reply:
done